pulsar-broker/src/main/java/org/apache/pulsar/broker/delayed/InMemoryDelayedDeliveryTracker.java

```diff
@@ -0,0 +1,193 @@
+
+    private final TripleLongPriorityQueue priorityQueue = new TripleLongPriorityQueue();
+
+    private final PersistentDispatcherMultipleConsumers dispatcher;
+
+    // Reference to the shared (per-broker) timer for delayed delivery
+    private final Timer timer;
+
+    // Current timeout or null if not set
+    private Timeout timeout;
+
+    // Timestamp at which the timeout is currently set
+    private long currentTimeoutTarget;
+
+    private final long tickTimeMillis;
+
+    private final Clock clock;
+
+    InMemoryDelayedDeliveryTracker(PersistentDispatcherMultipleConsumers dispatcher, Timer timer, long tickTimeMillis) {
+        this(dispatcher, timer, tickTimeMillis, Clock.systemUTC());
+    }
+
+    InMemoryDelayedDeliveryTracker(PersistentDispatcherMultipleConsumers dispatcher, Timer timer, long tickTimeMillis, Clock clock) {
+        this.dispatcher = dispatcher;
+        this.timer = timer;
+        this.tickTimeMillis = tickTimeMillis;
+        this.clock = clock;
+    }
+
+    @Override
+    public boolean addMessage(long ledgerId, long entryId, long deliveryAt) {
+        long now = clock.millis();
+        if (log.isDebugEnabled()) {
+            log.debug("[{}] Add message {}:{} -- Delivery in {} ms ", dispatcher.getName(), ledgerId, entryId,
+                    deliveryAt - now);
+        }
+        if (deliveryAt < now) {
+            // It's already about time to deliver this message
+            return false;
+        }
+
+        priorityQueue.add(deliveryAt, ledgerId, entryId);
+        updateTimer();
+        return true;
+    }
+
+    /**
+     * Return true if there's at least a message that is scheduled to be delivered already
+     */
+    @Override
+    public boolean hasMessageAvailable() {
+        return !priorityQueue.isEmpty() && priorityQueue.peekN1() <= clock.millis();
+    }
+
+    /**
+     * Get a set of position of messages that have already reached
+     */
+    @Override
+    public Set<PositionImpl> getScheduledMessages(int maxMessages) {
+        int n = maxMessages;
+        Set<PositionImpl> positions = new TreeSet<>();
+        long now = clock.millis();
+        // Pick all the messages that will be ready within the tick time period.
+        // This is to avoid keeping rescheduling the timer for each message at
+        // very short delay
+        long cutoffTime = now + tickTimeMillis;
+
+        while (n > 0 && !priorityQueue.isEmpty()) {
+            long timestamp = priorityQueue.peekN1();
+            if (timestamp > cutoffTime) {
+                break;
+            }
+
+            long ledgerId = priorityQueue.peekN2();
+            long entryId = priorityQueue.peekN3();
+            positions.add(new PositionImpl(ledgerId, entryId));
+
+            priorityQueue.pop();
+            --n;
+        }
+
+        if (log.isDebugEnabled()) {
+            log.debug("[{}] Get scheduled messags - found {}", dispatcher.getName(), positions.size());
+        }
+        updateTimer();
```

Comment:
why are we updating timer here?

Reply:
We took items out of the queue, we need to adjust the timer for the next scheduled message